Allow users to opt-out of welcome notifications

vaidehijoshi opened the pull request:
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [x] Optimization
- [ ] Documentation Update

## Description
Allows users to opt-out and unsubscribe from welcome notifications! I also refactored the tests and added some `TODO`s for @thepracticaldev/delightful.

Something to note here is that we will _always_ show the "unsubscribe" check box on the `/settings/notifications` page. I went with this approach because it seems weird to not show that option on the settings page (or even on the bottom of the notification) if someone's account was created 7 days ago (imagine someone logging in after 8 days and seeing these welcome notifications and wanting to unsubscribe but not seeing the option to do so?).

In the future, I'm sure we'll have more notifications sent via our platform, so we might want to consider splitting out the `/settings` route into `/settings/email` and `/settings/notifications`.

## Related Tickets & Documents
Resolves #6705.

## Mobile & Desktop Screenshots/Recordings (if there are UI changes)
<img width="1170" alt="Screen_Shot_2020-03-25_at_4_59_42_PM" src="https://user-images.githubusercontent.com/6921610/77596853-607d0580-6eba-11ea-85be-d20822b13364.png">
<img width="1168" alt="Screen_Shot_2020-03-25_at_4_59_51_PM" src="https://user-images.githubusercontent.com/6921610/77596857-6377f600-6eba-11ea-87f7-1cdec10bf89c.png">

## Added tests?

- [x] yes
- [ ] no, because they aren't needed
- [ ] no, because I need help

## Added to documentation?

- [ ] docs.dev.to
- [ ] readme
- [x] no documentation needed **but I added some `TODO`s for our team!**

rhymes commented on app/views/notifications/_broadcast.html.erb:
```diff
@@ -8,5 +8,9 @@
     <%= json_data["broadcast"]["processed_html"].html_safe %>
+
+    <% if notification.json_data['broadcast']['type_of'] == "Welcome" %>
+      <div class="opt-out"><a href="/settings/notifications">Go to your settings to manage your notification settings.</a></<div>
```
you could use `<%= user_settings_path(tab: :notifications) %>` here in lieu of the hardcoded path

vaidehijoshi replied:
ohhhhh duh of course, i should've remembered to do that 🙈

rhymes commented on app/views/users/_notifications.html.erb:
```diff
@@ -62,6 +62,16 @@
   </div>
+  <h3>Platform Notification Settings</h3>
+  <p>
+    <em>Notifications that only appear on the <a href="/notifications">notifications page</a>.</em>
```
you could use `<%= notifications_path %>` here in lieu of the hardcoded path

rhymes commented on spec/services/broadcasts/welcome_notification/generator_spec.rb:
```diff
@@ -24,32 +34,29 @@
+        expect(user.notifications.count).to eq(1)
       end
+
+      it_behaves_like "unsubscribed from welcome notifications"
```
what does this do? RSpec magic?

vaidehijoshi replied:
It applies [this shared example](https://github.com/thepracticaldev/dev.to/pull/6852/files#diff-133a9d8432d8554d80ce66e17cd9a94cR5-R11) in the context of this block :)

Basically my thought here is that a lot of this file could benefit from shared examples since we always want to test that we don't send any kind of welcome notification to a user who is unsubscribed, that we don't send double notifications to a user for any welcome notification, etc etc.

juliannatetreault replied:
I'm curious about this, too @rhymes ! 🧙

vaidehijoshi replied:
@juliannatetreault https://relishapp.com/rspec/rspec-core/docs/example-groups/shared-examples

mstruve commented on spec/requests/notifications_spec.rb:
```diff
@@ -538,15 +538,15 @@ def mock_heart_reaction_notifications(followers_amount, categories, reactable =
       end
 
-      it "renders the proper message" do
+      it "does not render the notification message" do
```
Nice catch!

mstruve commented on db/migrate/20200324170819_add_welcome_notifications_to_users.rb:
```diff
@@ -0,0 +1,5 @@
+class AddWelcomeNotificationsToUsers < ActiveRecord::Migration[5.2]
+  def change
+    add_column :users, :welcome_notifications, :boolean, default: true, null: false
```
`user.welcome_notifications` seems to me like it would return an active record relation of notifications rather than a boolean. What about something more boolean like `user.subscribed__to_welcome_notifications` or something

vaidehijoshi replied:
@mstruve I agree with you here, but all of the other boolean flags on the user are named the same way (`mod_roundrobin_notifications`, `email_follower_notifications`, etc.). It seemed better to follow the convention set rather than diverge from it here.

That being said, I do think we should rename all of these columns at some point (cool down period?), but it definitely feels out of scope for this PR.

mstruve replied:
Okie dokie!

vaidehijoshi replied:
@mstruve Created #6892 to make sure that this change doesn't get forgotten during the cool down period :)

juliannatetreault commented on app/views/users/_notifications.html.erb:
```diff
@@ -62,6 +62,16 @@
+  <div class="checkbox-field">
+    <div class="sub-field">
+      <%= f.check_box :welcome_notifications %>
+      <%= f.label :welcome_notifications, "Send me occasional tips and tricks" %>
```
Quick non-blocking nitpick: I feel like this copy here doesn't quite convey what the User is opting into or what Welcome Notifications are for. What if we had something like "Send me occasional tips on how to enhance my DEV experience" or something more along the lines of that?

vaidehijoshi replied:
Yeah, I don't love the copy I had and was hoping people would have ideas on something better. I can certainly update, but I'd also like @lisasy to weigh in here.

This whole page could generally be re-worked for copy, I think; the repeated "send me email" has been making my eyes twitch over here 👀

lisasy replied:
I like @juliannatetreault's revision to this — it's clear on the cadence and on what the tips will be for.

vaidehijoshi commented:
This should be good to go; fixed the failing spec spotted by @rhymes, 
addressed copy per @juliannatetreault + @lisasy. Will merge this once specs are green!

rhymes commented:
A small note: we should use `ApplicationConfig["COMMUNITY_NAME"]` for the messaging as hardcoding `DEV` will have an impact on other communities 😊

vaidehijoshi commented:
@rhymes OOPS, good catch. Addressed this in https://github.com/thepracticaldev/dev.to/pull/6893
